### ISSUE TITLE
Developing golang in nix. Advise needed

### ISSUE THREAD

**Issue by alex-glv** [NixOS/nixpkgs#13670]
I have gone through this guide: http://lethalman.blogspot.nl/2015/02/developing-in-golang-with-nix-package.html
Things mostly make sense, but I have no clue how to test my code. Especially, if I have subpackages. nix-build throws "src not set", so should I resort to running "go build" to run my app?
Second, if I define subPackages with my "innerpkg", running nix-build throws "can't load package: package github.com/myuser/project/innerpkg in any of: (lists GOPATH entries)"
I can build my project only with go build if I don't have any subpackages. If I don't have subpackages and I override phases with ' phases = [ "buildPhase" "installPhase"];' nix starts the build, but fails with the following:
```
building /nix/store/zn676by0m8f7z4js1cq4appx2j87gyqc-stdenv/setup: line 877: > pushd: go/src: No such file or directory
```
Can anyone suggest workflow? Thanks!

**Comment by heydojo** [comment]
Please use IRC or the mailing list if you have questions or need help.
The issue tracker here on github is for bug reports.
It is a little unclear if you are reporting a bug here.

**Comment by globin** [comment]
IMHO questions like this are valid, too, as they might point us to missing documentation :)

**Comment by alex-glv** [comment]
Maybe add a new tag: "documentation updates" or "discussion"?
I think having IRC guidance doesn't benefit much anyone but the one who asks for help on IRC.

**Comment by kamilchm** [comment]
I'm trying to do same thing. I got working ``nix-build`` with my [default.nix](https://github.com/kamilchm/go2nix/blob/master/default.nix) but I can't figure out how to use it with ``nix-shell``.
I want to be able to run commands like ``go test``, ``go build`` ... from ``nix-shell``.

**Comment by the-kenny** [comment]
What's the current status here? Can we close this?

**Comment by kamilchm** [comment]
After https://github.com/NixOS/nixpkgs/blob/9788dded2f05a51ecd60023c770c9dae857560d2/pkgs/development/go-modules/generic/default.nix#L190-L198 `buildGoPackage` is usable in `nix-shell` by default.
Do we need something more?

**Comment by pierrebeaucamp** [comment]
Yes, this is still not quite working yet for sub-packages.

Dependencies are correctly getting placed on `$GOPATH` (as per you snippet above), but the whole source folder is not. This means that you can't import a sub-package using the whole `goPackagePath`. Note that this works correctly for `nix-build`, only `nix-shell` is affected.

**Comment by kamilchm** [comment]
I think it's a overall good practice to place source of packages you're working on within the `GOPATH` tree and then use `nix` for dependencies only.
See other examples in tolls like https://github.com/GetStream/vg#usage

**Comment by pierrebeaucamp** [comment]
Sorry, I can't quite follow you. I'm aware that it is a best practice to have a folder structure like you described. However I thought that Nix would not require it, because it sets `$GOPATH` dynamically.

Moreover, nix already sets `$GOPATH` correctly when building it, so I think this is a genuine bug, not just some disagreement about folder structures. 

In my use case, I want to run nix on our CI system. Building our app is easy enough - `nix-build` does everything we need. But we would also like to run our tests separately, using `nix-shell --command "go test ./..."`, which fails because of this issue.

I'm aware that I can probably add a test phase to my `default.nix`, but it would be so much nicer to use nix-shell for both development and testing.

**Comment by pierrebeaucamp** [comment]
no, I did not try this yet, partly because I have not yet set up my computer to use a local version of `nixpkgs`. I will do that in the coming days - if this resolves my problem, I'd be happy to open a PR

**Comment by pierrebeaucamp** [comment]
So fyi I'm currently manually doing that symlinking and everything appears to be fine. I'm still somewhat new to nix, so I have no idea how to tell nix-shell to use a custom nixpkgs directory to test out your patch.

**Comment by tomberek** [comment]
close as answered?
